In [1]:
from copy import deepcopy
import matplotlib.pyplot as plt
import matplotlib as mpl
import numpy as np

%matplotlib widget

# Multivariate Gaussian

The equation for multivariate gaussian comes out as

$$
p(\mathbf{x}) = (2\pi)^{-d/2}|\Sigma|^{-1/2}
\exp^{\left( -\tfrac{1}{2} (\mathbf{x} - \boldsymbol{\mu})^\top 
\Sigma^{-1} (\mathbf{x} - \boldsymbol{\mu}) \right)}
$$

Components:
$d$ = dimension of the vector  
$x$ = random vector ($d×1$)  
$\mu$ = mean vector ($d×1$)  
$\Sigma$ = covariance matrix ($d×d$, symmetric & positive-definite)  
$|\Sigma|$= determinant of $\Sigma$  

## Implementation

In [6]:
# Provided values

measurements  = [1., 2., 3.]

x = np.array([0.,0.])[:,None] # Initial belief
P = np.array([[1000., 0.], [0., 1000.]]) # uncertainty
u = np.zeros_like(x) # external motion, here 0
F = np.array([[1., 1.], [0., 1.]]) # state transition matrix
H = np.array([[1., 0.]]) # measurement fn.
R = np.array([[1.]]) # measurement uncertainty
I = np.identity(2)

### Notes based on given values
x -> state (location and velocity)  
P -> no correlation between location and velocity, high uncertainty for both  
u -> is 0, so no effect  
F -> here, $x_{t+1} = x_t + \dot{x_t}$ , and $\dot{x_{t+1}} = \dot{x_t}$  
H -> here, only $x$ (location is measured)  
R -> location uncertainty, variance is 1, hence s.d is 1

In [7]:
def filter(x, P):
    """
        x -> state
        P -> covariance matrix (uncertainty)
    """
    for n in range(len(measurements)):
        z = measurements[n]

        # measurement update

        y = z - H @ x # error (residual)

        s = H @ P @ H.T  + R # projecting the error to measurement space??

        K = P @ H.T @ np.linalg.inv(s) # how much the estimate is to be moved from prediction.
        # K is a vector of  gains affecting each state variable
        # K=0, estimate = prediction, K = 1, estimate = measurement

        x = x + K * y # prediction + kalman gain * residual

        P = (I- K @ H) @ P

        # prediction update (for next time step) 

        x = F @ x + u

        P = F @ P @ F.T

        

        print(f"{x=}")
        print()
        print(f"{P=}")

In [8]:
filter(x, P)

x=array([[0.999001],
       [0.      ]])

P=array([[1000.999001, 1000.      ],
       [1000.      , 1000.      ]])
x=array([[2.99800299],
       [0.999002  ]])

P=array([[4.99002494, 2.99301795],
       [2.99301795, 1.99501297]])
x=array([[3.99966644],
       [0.99999983]])

P=array([[2.33189042, 0.99916761],
       [0.99916761, 0.49950058]])


**Things to Note**
  - The covariance matrix now has correlation (non-diagonal elements are larger numbers)
  - The variance (spread/confusion) decreased from the initial provided value 1000 to max of 2.33
  - The velocity is correctly estimated to be approx. 1
  - The filter is accurately predicting the next measurement

In [10]:
1-np.array([1])

array([0])